rfcs/0039-minaform-tool-design.md

```diff
@@ -0,0 +1,143 @@
+
+In addition to use on any developer laptop, the docker image could be deployed directly into the cluster, google cloud shell, or a google cloud VM to allow for low-latency deployment to gcloud even from a high-latency network (hope that helps @bkase @mrmr1993 !).
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Why Bash?
+
+* Key generation and upload are handled by bash scripts as of now so this will allow for quick iteration as hacky internals are removed, without changing the command line interface or the user experience. This design is intended to not introduce any additional tools to the process that we were not already committed to, so that after the minimum viable verison is in place we can re-evaluate if the system needs new tools.
+
+Why is this design the best in the space of possible designs?
+
+* To a point mentioned in the comments, bash has plenty of built-in support for properly parsing command-line arguments (see scripts/release-docker.sh in the mina repo for an example of handling that more properly), and very little needs to be done to glue terraform + coda-network together besides standardizing file paths and manipulating strings so the outputs of one tool properly match the inputs of another. A carefully written and commented bash script is the shortest path to a usable solution, and when the requirements for a solution exceed the plan outlined in this RFC we can just as easily continue from this point while rewriting in a compiled and typechecked language.
+
+What other designs have been considered and what is the rationale for not choosing them?
+
+* As I see it, the most reasonable non-bash language for this tooling is go, as the rest of the tooling used here (terraform, helm, prometheus, kubernetes, docker...) is available via go libraries, go modules can handle the versioning of all dependencies (except coda-network), and the infrastructure team is already well-versed in go and can iterate quickly with it. Why not start in go? because as of today the main dependencies of this tool include a hacky bash script and coda-network, which are not in go, and invoking go tools like terraform and kubectl are the easiest part of the process. Bash is the simplest direction to proceed from here, but when the complexity of this tool exceeds a reasonable amount for bash then a go rewrite is easy and viable. To rewrite in go from that point can rely on the bash implementation for A/B testing, and as the plan B if the more complex tool `go`es wrong.
```

Comment:
Haha while I'll take it hard to say that infra as a whole is well-versed in go and able to iterate quickly considering the lack of any Go tooling or src in our codebase (infra or otherwise) and compared to other languages like Python (used throughout `coda-automation` services and likely more familiar to new hires and probably most engineers nowadays). Also considering the eye towards usability and maintainability here, go while great for performance leaves quite a bit to be desired on the writeability + readability fronts; at least compared to langs like python, js, ocaml.